### PR TITLE
Prevent unexpected values from causing the websocket filtering to crash

### PR DESCRIPTION
**Pull request by jon-betts**
Turns out we do get more different values than the three expected, so
we need to account for this. Otherwise a single consumer can poison
the well.

This tackles this from both sides, but not adding anything we don't
expect, and also not accepting it even if it gets through.

### CODE REVIEW

File: h/streamer/filter.py [hunk 1]
```diff
@@ -52,9 +54,12 @@ def matching(cls, sockets, annotation):
...
             for field, value in socket.filter_rows:
-                if value in values[field]:
-                    yield socket
-                    break
+                try:
+                    if value in values[field]:
+                        yield socket
+                        break
+                except KeyError:
+                    continue
```
Comment (jon-betts) [inline h/streamer/filter.py:62]: Don't barf on bad fields

File: h/streamer/filter.py [hunk 2]
```diff
@@ -69,13 +74,15 @@ def set_filter(cls, socket, filter):
     def _rows_for(cls, filter):
         """Convert a filter to field value pairs."""
         for clause in filter["clauses"]:
+            field = clause["field"]
+            if field not in cls.KNOWN_FIELDS:
+                continue
+
```
Comment (jon-betts) [inline h/streamer/filter.py:80]: But also don't add them to the list in the first place.

File: tests/h/streamer/filter_test.py [hunk 3]
```diff
@@ -137,6 +169,10 @@ def get_randomized_filter(self):
             ],
         }
 
+    @pytest.fixture
+    def annotation(self, factories):
+        return factories.Annotation()
```
Comment (jon-betts) [inline tests/h/streamer/filter_test.py:174]: A few tests just needed a random annotation of some sort.

File: tests/h/streamer/filter_test.py [hunk 4]
```diff
@@ -45,29 +46,60 @@ def test_it_matches_uri(
...
+
+        filter_ = {
+            "match_policy": "include_any",
+            "actions": {},
+            "clauses": [{"field": field, "operator": "one_of", "value": value}],
+        }
+
+        SocketFilter.set_filter(socket, filter_)
+
+        assert socket.filter_rows == Any.iterable.containing(expected).only()
```
Comment (jon-betts) [inline tests/h/streamer/filter_test.py:101]: This was tested indirectly before in the fixture which added the rows and then tested them. This tests the behaviour explicitly, and gives us a place to check that the rows aren't added separately from testing that they don't effect matching.

File: tests/h/streamer/filter_test.py [hunk 5]
```diff
@@ -45,29 +46,60 @@ def test_it_matches_uri(
...
-        result = tuple(SocketFilter.matching([socket_no_rows], ann))
+        result = tuple(SocketFilter.matching([socket_no_rows], annotation))
         assert not result
 
+    def test_it_does_not_crash_with_unexpected_fields(self, annotation):
+        socket = FakeSocket()
+        socket.filter_rows = (("/not_a_thing", "value"),)
+
+        result = tuple(SocketFilter.matching([socket], annotation))
+        assert not result
```
Comment (jon-betts) [inline tests/h/streamer/filter_test.py:72]: This is the "we don't die" check

File: h/streamer/filter.py (hunk 1 above)
Comment (robertknight) [reply h/streamer/filter.py:62]: AIUI `filter_rows` _should_ only be set with `set_filter`, so any unexpected entries here would indicate an error in `set_filter`?

**Review by robertknight — APPROVED** [review]
LGTM. If I've understood correctly the check in `matching` is purely defensive and any unexpected entries that make their way into `filter_rows` would indicate an error in `set_filter`? If so I'd be inclined to raise an error here.

You mentioned in the PR description:

> Otherwise a single consumer can poison
the well.

Did the WebSocket server recover from the problem automatically if this happens? I presume it will crash and disconnect all existing clients? If the client sending problematic configuration auto-reconnects this could get into a loop.  It might be worth thinking about how to avoid this more generally - eg. if an exception occurs when dispatching a notification, just disconnecting the WebSocket involved.

File: h/streamer/filter.py (hunk 1 above)
Comment (jon-betts) [reply h/streamer/filter.py:62]: Yes... this should be safe for now, but saves us from a code change in `set_filter` causing a seemingly unrelated explosion.

**Comment by jon-betts** [comment]
I'm not really sure on the behavior, but if the client making the requests with a bad filter continues to make it they may just be evicted and go and cause trouble elsewhere. 

I don't think the whole socket crashes, just the filtering process, but it might as well have, as nobody gets any messages.

**Pull request merged**